src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs

```diff
@@ -551,6 +569,84 @@ public bool TryGetValue<TValue>(SyntaxTree tree, SyntaxTreeValueProvider<TValue>
         }
     }
 
+    /// <summary>
+    /// Context for a compilation unit start action.
+    /// A compilation unit start action can use a <see cref="CompilationUnitStartAnalysisContext"/> to register actions to be executed at any of:
+    /// <list type="bullet">
+    /// <item>
+    /// <description>start of semantic analysis of a method body or an expression appearing outside a method body,</description>
+    /// </item>
+    /// <item>
+    /// <description>completion of semantic analysis of a method body or an expression appearing outside a method body, or</description>
+    /// </item>
+    /// <item>
+    /// <description>completion of semantic analysis of a syntax node.</description>
+    /// </item>
+    /// </list>
+    /// </summary>
+    public abstract class CompilationUnitStartAnalysisContext
+    {
+        private readonly Compilation _compilation;
+        private readonly AnalyzerOptions _options;
+        private readonly CancellationToken _cancellationToken;
+        private readonly SyntaxTree _compilationUnit;
+
+        /// <summary>
+        /// <see cref="CodeAnalysis.Compilation"/> that is the subject of the analysis.
+        /// </summary>
+        public Compilation Compilation { get { return _compilation; } }
+
+        /// <summary>
+        /// Options specified for the analysis.
+        /// </summary>
+        public AnalyzerOptions Options { get { return _options; } }
+
+        /// <summary>
+        /// Token to check for requested cancellation of the analysis.
+        /// </summary>
+        public CancellationToken CancellationToken { get { return _cancellationToken; } }
+
+        public SyntaxTree CompilationUnit => _compilationUnit;
```

Comment:
Should this be exposing the SemanticModel instead or do we feel that the analyzers would perform purely syntactical checks in the start action?

Reply:
We discussed this in API review, and did not like the idea of exposing the semantic model.  we really wanted this for pre-checks before that.  if you want to do semnatic checks it would be through the callbacks you register.